src/TBuffer.cpp

```diff
@@ -650,17 +662,81 @@ void TBuffer::translateToPlainText(std::string& incoming, const bool isFromServe
+                    // (countermeasure against infinite recursion)
+                    TMxpProcessingResult const result =
+                            mpHost->mMxpProcessor.processMxpInput(ch, localBufferPosition >= endOfMXPEntity);
+
+                    switch (result) {
+                    case HANDLER_NEXT_CHAR:
                         localBufferPosition++;
                         continue;
-                    } else if (result == HANDLER_COMMIT_LINE) { // BR tag
+                    case HANDLER_COMMIT_LINE: // BR tag
                         ch = '\n';
                         goto COMMIT_LINE;
-                    } else { //HANDLER_FALL_THROUGH -> do nothing
+                    case HANDLER_INSERT_ENTITY_CUST:
+                        // custom entity value set with <!EN>, recurse except for other custom entities
+                    case HANDLER_INSERT_ENTITY_LIT: {
+                        // Unknown entity name like &unknown; push back into buffer for codeset interpretation,
+                        // but no MXP parsing.
+
+                        // We need to insert the entity value into the buffer at the current position to
+                        // process it
+                        size_t valueLength = mpHost->mMxpProcessor.getEntityValue().length();
+
+                        localBuffer.replace(0, localBufferPosition + 1, mpHost->mMxpProcessor.getEntityValue().toLatin1());
+
+                        if (result == HANDLER_INSERT_ENTITY_LIT) {
+                            if (localBufferPosition < endOfMXPEntity) {
+                                // This is a special case, our unknown entity might actually be a custom one
+                                // inside a custom one which we refused to resolve to avoid an endless recursion.
+                                // So we carefully adjust the end marker so custom entities are not reenabled to early
+                                endOfMXPEntity -= localBufferPosition + 1 - valueLength;
+                                endOfLiteralEntity = valueLength;
+                            } else {
+                                endOfMXPEntity = valueLength;
+                            }
+                            endOfLiteralEntity = valueLength;
+                        } else {
+                            // HANDLER_INSERT_ENTITY_CUST
+                            endOfMXPEntity = valueLength;
+                            endOfLiteralEntity = 0;
+                        }
+
+                        // Now restart the loop to parse the newly inserted text
+                        localBufferLength = localBuffer.length();
+                        localBufferPosition = 0;
+                        continue;
+                    }
+                    case HANDLER_INSERT_ENTITY_SYS: {
+                        // System entities are literal QString / UTF values which we just 'print'
```

Comment:
Exactly, this is one of the reason for the fuzz I make about localBuffer and distinguishing System/Default Entities (I'll talk about &newline; below): When a custom entity is defined like <!en custom "foobar"> the entity definition is made by the MXP handler before interpretation of the encoding, that is, the contents "foobar" is in whatever encoding used by the client (albeit internally stored in a QString).
Hence, when I put the contents now back into localBuffer, I reset localBufferPosition to ensure it is looped over again, now interpreting the encoding. It will also passed to the MXPProcessor again, as MXP protocol definition defines that an element like &lt;EM> might be part of the entity contents (this is the second reason). Through the use of the flags/pointers endOfMXPEntity and endOfLiteralEntity I ensure, that on the second pass: a) custom entities are no longer resolved (to avoid an endless loop caused by <!en loop "&loop;"> and b) we dont' get into an endless loop in case of an unknown entity where &unknown; resolves to &unknown;   endOfLiteralEntity signals that we are just looping over such an unknown entity name. It could be argued if this second loop over an unknown entity is required, but at least in theory it could contain characters in the clients encoding.

So, I think the encoding is handled find..

Reply:
I'll be honest I'm not sure I follow that completely. I suppose when an entity is defined in the incoming data stream it will be in the encoding the server is currently using - so it is reasonable to replace an "entity" with the bytes (can be stored in `QByteArray`s rather than `QString`s) that that represents.

To be honest though I do not think Zugg ever gave a thought to anything other than English (as ASCII) as the data stream - other than providing for *named* "entities" to provide a subset. Does MXP do anything with `&#`...*hexDigits*...`;` or `&`...*decimalDigits*...`;` "entities"?

Reply:
MXP does not say anything about &decimal or &#hex . I'd tend to agree with you, he only had US-Ascii in mind with a few special chars. That said, the mudlet entity resolver already does &#hex. It was this way, didn't touch this part. &decimal is not understood.. looking at the source (TEntityResolver:62) one could think of this as a bug. one-liner to fix, but I had it not on my screen and it was no priority, as MXP does not define this. Maybe as a separate PR. Also, for my personal taste, the code should check the entity consist of valid hex/decimal digits b4 squeezing this into a QChar/QString.